packages/botonic-plugin-contentful/tests/util/memoizer.test.ts

```diff
@@ -0,0 +1,172 @@
+import {
+  cacheForeverStrategy,
+  fallbackStrategy,
+  Memoizer,
+} from '../../src/util/memoizer'
+
+class MockF {
+  callsFA = 0
+  callsFB = 0
+  error: Error | undefined
+  salt = ''
+
+  fA(a: number, b: string): Promise<string> {
+    this.callsFA++
+    if (this.error) {
+      return Promise.reject(this.error)
+    }
+    return Promise.resolve('A' + String(a) + b + this.salt)
+  }
+
+  fB(a = 1, b = 'b'): Promise<string> {
+    this.callsFB++
+    if (this.error) {
+      return Promise.reject(this.error)
+    }
+    return Promise.resolve('B' + String(a) + b + this.salt)
+  }
+}
+
+describe('Memoizer', () => {
+  it('TEST: cacheForeverStrategy common properties', async () => {
+    await assertCommonStrategyProperties(
+      () => new Memoizer(cacheForeverStrategy),
+      false
+    )
+  })
+  it('TEST: cacheForeverStrategy only fails if last call failed', async () => {
+    const sut = new Memoizer(cacheForeverStrategy)
+    const mock = new MockF()
+    const memoized = sut.memoize(mock.fA.bind(mock))
+
+    // act/assert
+    mock.error = new Error('forced failure')
+    await expect(memoized(2, 'b')).rejects.toThrowError(mock.error)
+    mock.error = undefined
+    await expect(memoized(2, 'b')).resolves.toBe('A2b')
+    expect(mock.callsFA).toBe(2)
+  })
+
+  function usingFallback(funcName: string, args: any[], e: any): Promise<void> {
+    console.error(
+      `Using fallback for ${funcName}(${String(args)}) after error: ${String(
+        e
+      )}`
+    )
+    return Promise.resolve()
+  }
+
+  it('TEST: fallbackStrategy common properties', async () => {
+    await assertCommonStrategyProperties(
+      () => new Memoizer(fallbackStrategy(usingFallback)),
+      true
+    )
+  })
+
+  it('TEST: fallbackStrategy uses last invocation result', async () => {
+    const sut = new Memoizer(fallbackStrategy(usingFallback))
+    const mock = new MockF()
+    const memoized = sut.memoize(mock.fA.bind(mock))
+
+    // act/assert
+    await expect(memoized(2, 'b')).resolves.toBe('A2b')
+    mock.salt = 'pepper'
+    await expect(memoized(2, 'b')).resolves.toBe('A2b' + mock.salt)
+    expect(mock.callsFA).toBe(2)
+  })
+
+  test('TEST: fallbackStrategy returns latest success return if function fails', async () => {
+    const sut = new Memoizer(fallbackStrategy(usingFallback))
+    const mock = new MockF()
+    const memoized = sut.memoize(mock.fA.bind(mock))
+
+    // arrange
+    await memoized(2, 'b')
+    mock.salt = 'pepper'
+    await memoized(2, 'b')
+
+    mock.error = new Error('forced failure')
+    await memoized(2, 'b')
+
+    // act
+    mock.error = undefined
+    await expect(memoized(2, 'b')).resolves.toBe('A2bpepper')
+    expect(mock.callsFA).toBe(4)
```

Comment:
Why `mock.callsFa` has to be called 4 times, wouldn't the last 2 times return the cached value instead of calling the `fA` function?

Reply:
This cache is not a "typical cache", since it *always* perform the call. It only returns a cached value when the last invocation failed.